commit 7f46252ff5317a97b4f6a3073c95ea7b5a87f1f5 (refs/changes/61/5161/2)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-11-29T09:58:13-08:00 (10 months ago)
    
    TRITON-989 improve sdcadm logging
    Reviewed by: Pedro Palazón Candel <pedro@joyent.com>
    Approved by: Pedro Palazón Candel <pedro@joyent.com>

diff --git a/CHANGES.md b/CHANGES.md
index 16b97b7..887e33a 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,12 @@
 
 # sdcadm Changelog
 
+## 1.23.0
+
+- TRITON-889 Improvements to sdcadm's trace logging to file.
+- TRITON-990 Change 'sdcadm check-health' and 'sdcadm check-config' to *no
+  longer* write a trace log to file.
+
 ## 1.22.0
 
 - TRITON-987 fabric setup broken by needless dependency on broken CA
diff --git a/lib/channel.js b/lib/channel.js
index 6d47e7b..1e5606a 100644
--- a/lib/channel.js
+++ b/lib/channel.js
@@ -238,6 +238,8 @@ ChannelCLI.prototype.do_set.options = [
     }
 ];
 
+ChannelCLI.prototype.do_set.logToFile = true;
+
 
 ChannelCLI.prototype.do_unset =
 function do_unset(subcmd, opts, _args, cb) {
@@ -290,6 +292,8 @@ ChannelCLI.prototype.do_unset.options = [
 ];
 
 ChannelCLI.prototype.do_unset.hidden = true;
+ChannelCLI.prototype.do_unset.logToFile = true;
+
 
 ChannelCLI.prototype.do_get = function do_get(subcmd, opts, _args, cb) {
     var self = this;
diff --git a/lib/cli/do_add_new_agent_svcs.js b/lib/cli/do_add_new_agent_svcs.js
index fba2923..9aaf65b 100644
--- a/lib/cli/do_add_new_agent_svcs.js
+++ b/lib/cli/do_add_new_agent_svcs.js
@@ -42,8 +42,10 @@ do_add_new_agent_svcs.help = [
     '',
     '{{options}}'
 ].join('\n');
+
 do_add_new_agent_svcs.hidden = true;
 
+do_add_new_agent_svcs.logToFile = true;
 
 // --- exports
 
diff --git a/lib/cli/do_fix_core_vm_resolvers.js b/lib/cli/do_fix_core_vm_resolvers.js
index 040915c..5db6af8 100644
--- a/lib/cli/do_fix_core_vm_resolvers.js
+++ b/lib/cli/do_fix_core_vm_resolvers.js
@@ -90,6 +90,8 @@ do_fix_core_vm_resolvers.help = (
     'that are not on the admin network.\n'
 );
 
+do_fix_core_vm_resolvers.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/do_install_docker_cert.js b/lib/cli/do_install_docker_cert.js
index c17c921..89b575e 100644
--- a/lib/cli/do_install_docker_cert.js
+++ b/lib/cli/do_install_docker_cert.js
@@ -155,6 +155,8 @@ do_install_docker_cert.help = (
     '{{options}}'
 );
 
+do_install_docker_cert.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/do_nfs_volumes.js b/lib/cli/do_nfs_volumes.js
index 1b27eeb..bc42aa3 100644
--- a/lib/cli/do_nfs_volumes.js
+++ b/lib/cli/do_nfs_volumes.js
@@ -653,6 +653,9 @@ do_nfs_volumes.help = (
     '{{options}}'
 );
 
+do_nfs_volumes.logToFile = true;
+
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/do_update_agents.js b/lib/cli/do_update_agents.js
index df8e90e..f1fcae0 100644
--- a/lib/cli/do_update_agents.js
+++ b/lib/cli/do_update_agents.js
@@ -1066,6 +1066,8 @@ do_update_agents.options = [
     }
 ];
 
+do_update_agents.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/do_update_docker.js b/lib/cli/do_update_docker.js
index f62f022..0ecaf0c 100644
--- a/lib/cli/do_update_docker.js
+++ b/lib/cli/do_update_docker.js
@@ -55,6 +55,8 @@ do_update_docker.options = [
 do_update_docker.help =
         'This command has been replaced by `sdcadm post-setup docker`.';
 
+do_update_docker.logToFile = true;
+
 module.exports = {
     do_update_docker: do_update_docker
 };
diff --git a/lib/cli/do_update_gz_tools.js b/lib/cli/do_update_gz_tools.js
index f9f0033..bf93bd4 100644
--- a/lib/cli/do_update_gz_tools.js
+++ b/lib/cli/do_update_gz_tools.js
@@ -899,6 +899,9 @@ do_update_gz_tools.help = (
     '{{options}}'
 );
 
+do_update_gz_tools.logToFile = true;
+
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/do_update_other.js b/lib/cli/do_update_other.js
index 16e74bd..479463b 100644
--- a/lib/cli/do_update_other.js
+++ b/lib/cli/do_update_other.js
@@ -909,6 +909,8 @@ do_update_other.help = (
     '{{options}}'
 );
 
+do_update_other.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/cli/experimental.js b/lib/cli/experimental.js
index ecfcc5a..111dbf7 100644
--- a/lib/cli/experimental.js
+++ b/lib/cli/experimental.js
@@ -50,6 +50,9 @@ ExperimentalCLI.prototype.init = function init(_opts, _args, _callback) {
     Cmdln.prototype.init.apply(this, arguments);
 };
 
+ExperimentalCLI.hidden = true;
+
+
 ExperimentalCLI.prototype.do_info = require('./do_info').do_info;
 
 
diff --git a/lib/cli/index.js b/lib/cli/index.js
index f7cb0cf..c46e0fc 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -93,21 +93,43 @@ CLI.prototype.init = function init(opts, args, callback) {
 
     // Generate a UUID we can use both logs:
     this.uuid = uuid();
+
     // Setup the logger.
-    var handler = this.handlerFromSubcmd(args[0]);
-    var logComponent = args[0] || 'nosubcmd';
-    if (handler && handler.name && handler.name.slice(0, 3) === 'do_') {
-        // Use this to canonicalize the name, e.g. if args[0] is 'insts',
-        // but the command is canonically 'instances'.
-        logComponent = handler.name.slice(3);
+    //
+    // First look for a `logToFile` boolean on anything in the CLI handler
+    // chain. If found, we do trace-level logging to
+    // "/var/log/sdcadm/logs/*.log".
+    //
+    // Dev Note: this `handlerFromSubcmd` walking process isn't perfect if
+    // there are interspersed options, e.g. `sdcadm --some-opt blah subcmd ...`.
+    // The only fall out is that the "logComponent" part of the temporary
+    // log file in /var/log/sdcadm/logs won't have the complete subcommand
+    // name.
+    var logToFile = false;
+    var argIdx = 0;
+    var handler;
+    while (argIdx < args.length) {
+        var nextHandler = (handler || this).handlerFromSubcmd(args[argIdx]);
+        if (!nextHandler) {
+            break;
+        }
+        handler = nextHandler;
+        if (handler.hasOwnProperty('logToFile')) {
+            logToFile = handler.logToFile;
+            break;
+        }
+        if (typeof (handler.handlerFromSubcmd) !== 'function') {
+            break;
+        }
+        argIdx++;
     }
-    var logToFile = (handler && handler.logToFile || false);
-    var skipInit = false;
+    var logComponent = args.slice(0, argIdx + 1).join('-');
 
-    if (args.indexOf('help') !== -1 ||
-        args.indexOf('--help') !== -1 ||
-        args.indexOf('-h') !== -1 ||
-        opts.help) {
+    // Try to guess if this is just *--help* output. If so, skip trace logging
+    // to file and other expensive init work.
+    var skipInit = false;
+    var lastArg = (args.length ? args[args.length - 1] : null);
+    if (lastArg === '--help' || lastArg === '-h') {
         logToFile = false;
         skipInit = true;
     }
@@ -221,8 +243,12 @@ CLI.prototype.fini = function fini(subcmd, err, cb) {
         if (err && this.opts && this.opts.verbose) {
             logLevel = 'error';
         }
-        this.log[logLevel]({subcmd: subcmd, exitStatus: exitStatus, cli: true},
-            'cli exit');
+        this.log[logLevel]({
+            err: err,
+            subcmd: subcmd,
+            exitStatus: exitStatus,
+            cli: true
+        }, 'cli exit');
     }
 
     cb();
@@ -440,7 +466,6 @@ CLI.prototype.do_check_config = require('./do_check_config').do_check_config;
 CLI.prototype.do_check_health = require('./do_check_health').do_check_health;
 
 CLI.prototype.do_dc_maint = DCMaintCLI;
-CLI.prototype.do_dc_maint.logToFile = true;
 
 
 experimental.ExperimentalCLI.prototype.do_avail =
@@ -449,24 +474,16 @@ available.do_experimental_avail;
 experimental.ExperimentalCLI.prototype.do_update =
 update.do_experimental_update;
 
-
 CLI.prototype.do_experimental = experimental.ExperimentalCLI;
-CLI.prototype.do_experimental.hidden = true;
-CLI.prototype.do_experimental.logToFile = true;
-
 
 
 CLI.prototype.do_post_setup = PostSetupCLI;
-CLI.prototype.do_post_setup.logToFile = true;
 
 CLI.prototype.do_platform = PlatformCLI;
-CLI.prototype.do_platform.logToFile = true;
 
 CLI.prototype.do_channel = ChannelCLI;
-CLI.prototype.do_channel.logToFile = true;
 
 CLI.prototype.do_default_fabric = defFabric.do_default_fabric;
-CLI.prototype.do_default_fabric.logToFile = true;
 
 CLI.prototype.do_completion = require('./do_completion');
 
diff --git a/lib/dc-maint.js b/lib/dc-maint.js
index 97bd189..bb7ab9c 100644
--- a/lib/dc-maint.js
+++ b/lib/dc-maint.js
@@ -193,6 +193,8 @@ DCMaintCLI.prototype.do_start.options = [
     }
 ];
 
+DCMaintCLI.prototype.do_start.logToFile = true;
+
 
 DCMaintCLI.prototype.do_stop = function do_stop(subcmd, opts, _args, cb) {
     var self = this;
@@ -228,6 +230,7 @@ DCMaintCLI.prototype.do_stop.help = (
         '\n' +
         '{{options}}'
 );
+DCMaintCLI.prototype.do_stop.logToFile = true;
 
 DCMaintCLI.prototype.do_status = function do_status(subcmd, opts, _args, cb) {
     var self = this;
@@ -295,6 +298,7 @@ DCMaintCLI.prototype.do_status.help = (
         '{{options}}'
 );
 
+
 // --- exports
 
 module.exports = {
diff --git a/lib/default-fabric.js b/lib/default-fabric.js
index 55c6838..8826583 100644
--- a/lib/default-fabric.js
+++ b/lib/default-fabric.js
@@ -183,6 +183,8 @@ do_default_fabric.options = [
     }
 ];
 
+do_default_fabric.logToFile = true;
+
 
 module.exports = {
     do_default_fabric: do_default_fabric,
diff --git a/lib/platform.js b/lib/platform.js
index 28e3420..3823ed6 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1690,6 +1690,7 @@ PlatformCLI.prototype.do_install.options = [
     }
 ];
 
+PlatformCLI.prototype.do_install.logToFile = true;
 
 
 /*
@@ -1811,6 +1812,8 @@ PlatformCLI.prototype.do_assign.options = [
     }
 ];
 
+PlatformCLI.prototype.do_assign.logToFile = true;
+
 
 PlatformCLI.prototype.do_list =
 function do_list(subcmd, opts, _args, cb) {
@@ -2189,6 +2192,8 @@ PlatformCLI.prototype.do_remove.options = [
     }
 ];
 
+PlatformCLI.prototype.do_remove.logToFile = true;
+
 
 PlatformCLI.prototype.do_avail = function do_avail(subcmd, opts, _args, cb) {
     var self = this;
@@ -2378,6 +2383,9 @@ PlatformCLI.prototype.do_set_default.options = [
     }
 ];
 
+PlatformCLI.prototype.do_set_default.logToFile = true;
+
+
 // --- exports
 
 module.exports = {
diff --git a/lib/post-setup/cloudapi.js b/lib/post-setup/cloudapi.js
index 586580f..273638e 100644
--- a/lib/post-setup/cloudapi.js
+++ b/lib/post-setup/cloudapi.js
@@ -201,6 +201,7 @@ do_cloudapi.options = [
     }
 ];
 
+do_cloudapi.logToFile = true;
 
 
 // --- exports
diff --git a/lib/post-setup/cmon.js b/lib/post-setup/cmon.js
index 14cfbdb..414348b 100644
--- a/lib/post-setup/cmon.js
+++ b/lib/post-setup/cmon.js
@@ -599,6 +599,9 @@ do_cmon.help = (
     '{{options}}'
 );
 
+do_cmon.logToFile = true;
+
+
 // --- exports
 
 module.exports = {
diff --git a/lib/post-setup/cns.js b/lib/post-setup/cns.js
index 7c6ff42..8af18f5 100644
--- a/lib/post-setup/cns.js
+++ b/lib/post-setup/cns.js
@@ -278,6 +278,8 @@ do_cns.help = (
     '{{options}}'
 );
 
+do_cns.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/post-setup/common-external-nics.js b/lib/post-setup/common-external-nics.js
index ea615b1..263a5d1 100644
--- a/lib/post-setup/common-external-nics.js
+++ b/lib/post-setup/common-external-nics.js
@@ -77,6 +77,8 @@ do_common_external_nics.options = [
     }
 ];
 
+do_common_external_nics.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/post-setup/dev-headnode-prov.js b/lib/post-setup/dev-headnode-prov.js
index 0dbff7c..22a855e 100644
--- a/lib/post-setup/dev-headnode-prov.js
+++ b/lib/post-setup/dev-headnode-prov.js
@@ -149,7 +149,7 @@ do_dev_headnode_prov.options = [
     }
 ];
 
-
+do_dev_headnode_prov.logToFile = true;
 
 // --- exports
 
diff --git a/lib/post-setup/dev-sample-data.js b/lib/post-setup/dev-sample-data.js
index e9c192f..660d00c 100644
--- a/lib/post-setup/dev-sample-data.js
+++ b/lib/post-setup/dev-sample-data.js
@@ -190,6 +190,7 @@ do_dev_sample_data.options = [
     }
 ];
 
+do_dev_sample_data.logToFile = true;
 
 
 // --- exports
diff --git a/lib/post-setup/docker.js b/lib/post-setup/docker.js
index 158f043..ae694c7 100644
--- a/lib/post-setup/docker.js
+++ b/lib/post-setup/docker.js
@@ -852,6 +852,8 @@ do_docker.help = [
     ''
 ].join('\n');
 
+do_docker.logToFile = true;
+
 // --- exports
 
 module.exports = {
diff --git a/lib/post-setup/fabrics.js b/lib/post-setup/fabrics.js
index 5224d6d..e136736 100644
--- a/lib/post-setup/fabrics.js
+++ b/lib/post-setup/fabrics.js
@@ -908,6 +908,7 @@ do_fabrics.options = [
     }
 ];
 
+do_fabrics.logToFile = true;
 
 module.exports = {
     do_fabrics: do_fabrics
diff --git a/lib/post-setup/ha-binder.js b/lib/post-setup/ha-binder.js
index 9dec9e2..f6f5371 100644
--- a/lib/post-setup/ha-binder.js
+++ b/lib/post-setup/ha-binder.js
@@ -795,7 +795,7 @@ do_ha_binder.help = (
     '\n'
 );
 
-
+do_ha_binder.logToFile = true;
 
 
 // --- exports
diff --git a/lib/post-setup/ha-manatee.js b/lib/post-setup/ha-manatee.js
index 5e1d124..73b2308 100644
--- a/lib/post-setup/ha-manatee.js
+++ b/lib/post-setup/ha-manatee.js
@@ -851,7 +851,7 @@ do_ha_manatee.help = (
     '{{options}}'
 );
 
-
+do_ha_manatee.logToFile = true;
 
 // --- exports
 
diff --git a/lib/post-setup/underlay-nics.js b/lib/post-setup/underlay-nics.js
index c230b10..ed4373f 100644
--- a/lib/post-setup/underlay-nics.js
+++ b/lib/post-setup/underlay-nics.js
@@ -371,6 +371,7 @@ do_underlay_nics.options = [
     }
 ];
 
+do_underlay_nics.logToFile = true;
 
 // --- exports
 
diff --git a/lib/post-setup/volapi.js b/lib/post-setup/volapi.js
index 2172b97..5a19bb2 100644
--- a/lib/post-setup/volapi.js
+++ b/lib/post-setup/volapi.js
@@ -508,6 +508,9 @@ do_volapi.help = (
     '{{options}}'
 );
 
+do_volapi.logToFile = true;
+
+
 // --- exports
 
 module.exports = {
diff --git a/package.json b/package.json
index 29eab86..73e08a1 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdcadm",
   "description": "Administer a SmartDataCenter (SDC) standup",
-  "version": "1.22.0",
+  "version": "1.23.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
